python/lsst/rubintv/analysis/service/command.py

```diff
@@ -0,0 +1,255 @@
+    response_type :
+        The type of response that this command sends to the user.
+        This should be unique for each command.
+    """
+
+    result: dict | None = None
+    response_type: str
+
+    @abstractmethod
+    def build_contents(self, databases: dict[str, DatabaseConnection], butler: Butler | None) -> dict:
+        """Build the contents of the command.
+
+        Parameters
+        ----------
+        databases :
+            The database connections.
+        butler :
+            A conencted Butler.
+
+        Returns
+        -------
+        contents :
+            The contents of the response to the user.
+        """
+        pass
+
+    def execute(self, databases: dict[str, DatabaseConnection], butler: Butler | None):
+        """Execute the command.
+
+        This method does not return anything, buts sets the `result`,
+        the JSON formatted string that is sent to the user.
+
+        Parameters
+        ----------
+        databases :
+            The database connections.
+        butler :
+            A conencted Butler.
+
+        """
+        self.result = {"type": self.response_type, "content": self.build_contents(databases, butler)}
+
+    def to_json(self):
+        """Convert the `result` into JSON."""
+        if self.result is None:
+            raise CommandExecutionError(f"Null result for command {self.__class__.__name__}")
+        return json.dumps(self.result)
+
+    @classmethod
+    def register(cls, name: str):
+        """Register a command."""
+        command_registry[name] = cls
+
+
+# Registry of all commands
+command_registry = {}
+
+
+def execute_command(command_str: str, databases: dict[str, DatabaseConnection], butler: Butler | None) -> str:
+    """Parse a JSON formatted string into a command and execute it.
+
+    Command format:
+    ```
+    {
+        name: command name,
+        content: command content (usually a dict)
+    }
+    ```
+
+    Parameters
+    ----------
+    command_str :
+        The JSON formatted command received from the user.
+    databases :
+        The database connections.
+    butler :
+        A conencted Butler.
+    """
+    try:
+        command_dict = json.loads(command_str)
+        if not isinstance(command_dict, dict):
+            raise CommandParsingError(f"Could not generate a valid command from {command_str}")
+    except Exception as err:
+        logging.exception("Error converting command to JSON.")
+        return error_msg(err)
+
+    try:
+        if "name" not in command_dict.keys():
+            raise CommandParsingError("No command 'name' given")
+
+        if command_dict["name"] not in command_registry.keys():
+            raise CommandParsingError(f"Unrecognized command '{command_dict['name']}'")
+
+        if "parameters" in command_dict:
+            parameters = command_dict["parameters"]
+        else:
+            parameters = {}
+
+        command = command_registry[command_dict["name"]](**parameters)
```

Comment:
I did it this way on purpose, although we can discuss the merits of my thinking, to solve two different problems.

The first purpose it serves is to make it easy for a developer to construct commands on the client and ensure that their structure matches the expected parameters of the class. The command send from the client is always going to be a json formatted string, and inspecting the class' attributes gives developers a quick way to ensure that the command string contains all of the required parameters of the command. I thought that this was cleaner than having to inspect the input parameters of the `build_contents` method. On a side note, I've never liked how python allows a class method to be overridden with required parameters that are different from the base class. In more strongly typed languages this is forbidden and I think that python allowing it can lead to difficult bugs to track, so I prefer to following a more stringent style of inheritance.

The second purpose is to separate the parsing of a command from the execution of a command. By structuring it this way I can separately check that a command can be initialized (parsing) and that the command executes correctly. I _could_ also achieve this by checking for a `TypeError` when the command method is executed, but then a type error that occurs somewhere else in the command could get confused with a type error that occurs when creating a command, so this seems cleaner.

Reply:
I just remarked this because this way you will have to pass all the dependencies on the execute method and they may not need them (Some Commands may need the Butler or the DB, but not both of them), if dependencies keep on growing then would be strange to call execute with arguments that are not using. The natural way to build  instances of a class is to pass the dependencies of the class when instantiated (Butler, DB connections...) and on each method pass the parameters they need to be executed.

Reply:
Just wanted to point that (and the way I use to solve this situation).